### PR TITLE
Add dedicated ServiceAccount for perftune jobs

### LINKED ISSUE #1975: Perftune jobs should use dedicated serviceaccount
At this point, perftune jobs that run in `scylla-operator-node-tuning` namespace use the default ServiceAccount which makes it impossible to change the individual permissions of perftune jobs that need to run privileged operations. We should wire up a dedicated service account for them which will also unblock running them on OpenShift where they need to be granted additional RBAC rules.

### PR DESCRIPTION
**Pull request by tnozicka**
**Description of your changes:**
This PR adds a dedicated ServiceAccount to use for perftune jobs which enhances security isolation and allows some platforms (like OpenShift) to grant it extra permissions.

**Which issue is resolved by this Pull Request:**
Resolves #1975

### Requires:
- [x] https://github.com/scylladb/scylla-operator/pull/2000

### CODE REVIEW

**Comment by tnozicka** [comment]
@zimnx @rzetelskik ptal

File: pkg/controller/nodeconfig/sync.go [hunk 1]
```diff
@@ -73,6 +78,11 @@ func (ncc *Controller) sync(ctx context.Context, key string) error {
 		objectErrs = append(objectErrs, err)
 	}
 
+	roleBindings, err := ncc.getRoleBindings()
```
Comment (zimnx) [inline pkg/controller/nodeconfig/sync.go:81]: This file already has a mixed way of getting dependent objects, lets use new approach for new code.

File: pkg/controller/nodeconfig/sync_rolebindings.go [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+	return utilerrors.NewAggregate(errs)
+}
+
+func (ncc *Controller) syncRoleBindings(
+	ctx context.Context,
+	roleBindings map[string]*rbacv1.RoleBinding,
+) error {
+	requiredRoleBindings := ncc.makeRoleBindings()
+
+	// Delete any excessive RoleBindings.
```
Comment (rzetelskik) [inline pkg/controller/nodeconfig/sync_rolebindings.go:62]: nit: redundant comment imo

File: pkg/controller/nodeconfig/sync.go (hunk 1 above)
Comment (rzetelskik) [reply pkg/controller/nodeconfig/sync.go:81]: doesn't `controllerhelpers.GetObjects` only apply to controlled objects, while these are non-namespaced? Unless you mean something else by new approach

Comment (zimnx) [reply pkg/controller/nodeconfig/sync.go:81]: They are originating from NodeConfig so there should be a controllerRef on them, and these are namespaced RoleBindings and Roles.

File: pkg/controller/nodeconfig/sync_rolebindings.go [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+	// Delete has to be the first action to avoid getting stuck on quota.
+	if err := ncc.pruneRoleBindings(ctx, requiredRoleBindings, roleBindings); err != nil {
+		return fmt.Errorf("can't delete RoleBinding(s): %w", err)
+	}
```
[the remark this answers is not in the document]
Comment (tnozicka) [reply pkg/controller/nodeconfig/sync_rolebindings.go:66]: copied it from the clusterrole part, but lets do this

File: pkg/controller/nodeconfig/sync.go (hunk 1 above)
Comment (tnozicka) [reply pkg/controller/nodeconfig/sync.go:81]: the existing SA doesn't use controller ref either

https://github.com/scylladb/scylla-operator/blob/11394625d288542782cd23fee50a3d9ff6957ebe/pkg/controller/nodeconfig/resource.go#L29-L39
These are shared by multiple NodeConfig and a for the lack of a better place we reconcile them with each of them.

File: pkg/controller/nodeconfig/sync_rolebindings.go (hunk 2 above)
Comment (tnozicka) [reply pkg/controller/nodeconfig/sync_rolebindings.go:25]: it's in all the other functions, it this one worth the inconsistency?

File: pkg/controller/nodeconfig/sync_rolebindings.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (tnozicka) [reply pkg/controller/nodeconfig/sync_rolebindings.go]: done

File: pkg/controller/nodeconfig/sync_rolebindings.go (hunk 2 above)
Comment (rzetelskik) [reply pkg/controller/nodeconfig/sync_rolebindings.go:25]: doesn't matter that much, hence a nit. Feel free to ignore if you prefer to keep it consistent

File: pkg/controller/nodeconfig/sync_roles.go [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+}
+
+func (ncc *Controller) syncRoles(ctx context.Context, Roles map[string]*rbacv1.Role) error {
+	requiredRoles := ncc.makeRoles()
+
+	// Delete any excessive Roles.
+	// Delete has to be the first action to avoid getting stuck on quota.
+	if err := ncc.pruneRoles(ctx, requiredRoles, Roles); err != nil {
+		return fmt.Errorf("can't delete Role(s): %w", err)
+	}
```
[the remark this answers is not in the document]
Comment (tnozicka) [reply pkg/controller/nodeconfig/sync_roles.go]: done

File: pkg/controller/nodeconfig/sync_roles.go [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+	"context"
+	"fmt"
+
+	"github.com/scylladb/scylla-operator/pkg/resourceapply"
+	rbacv1 "k8s.io/api/rbac/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	utilerrors "k8s.io/apimachinery/pkg/util/errors"
+)
+
+func (ncc *Controller) makeRoles() []*rbacv1.Role {
```
[the remark this answers is not in the document]
Comment (tnozicka) [reply pkg/controller/nodeconfig/sync_roles.go]: inlined

File: pkg/controller/nodeconfig/sync_roles.go [hunk 6]
```diff
@@ -0,0 +1,77 @@
...
+	Roles := []*rbacv1.Role{
```
[the remark this answers is not in the document]
Comment (tnozicka) [reply pkg/controller/nodeconfig/sync_roles.go]: fixed

File: pkg/resourceapply/rbac_test.go [hunk 7]
```diff
@@ -1565,3 +1565,492 @@ func TestApplyRoleBinding(t *testing.T) {
...
+
+						for i := range roleList.Items {
+							err := roleCache.Add(&roleList.Items[i])
+							if err != nil {
+								t.Fatal(err)
+							}
+						}
+					}
+
+					gotSts, gotChanged, gotErr := ApplyRole(ctx, client.RbacV1(), roleLister, recorder, tc.required, ApplyOptions{})
```
Comment (zimnx) [inline pkg/resourceapply/rbac_test.go:2008]: s/Sts/Role here and below

File: pkg/controller/nodeconfig/sync.go (hunk 1 above)
Comment (zimnx) [reply pkg/controller/nodeconfig/sync.go:81]: Then enqueueing owner in event handlers must not work and your PR does exactly that

File: pkg/controller/nodeconfig/sync_roles.go [hunk 8]
```diff
@@ -0,0 +1,47 @@
...
+)
+
+func (ncc *Controller) syncRoles(ctx context.Context, nc *scyllav1alpha1.NodeConfig, roles map[string]*rbacv1.Role) error {
+	requiredRoles := []*rbacv1.Role{
+		makePerftuneRole(),
+	}
+
+	// Delete any excessive Roles.
+	// Delete has to be the first action to avoid getting stuck on quota.
+	err := controllerhelpers.Prune(
```
Comment (zimnx) [inline pkg/controller/nodeconfig/sync_roles.go:23]: shouldn't Prune complain about missing controllerref on deleted objects? We are not allowed to delete them. There should be an option to disable it on justified cases similar to apply functions.

File: pkg/controller/nodeconfig/sync.go (hunk 1 above)
Comment (tnozicka) [reply pkg/controller/nodeconfig/sync.go:81]: yeah, that should enqueueAll but it must be broken in all the other places already there that I copied it from

File: pkg/controller/nodeconfig/sync_roles.go (hunk 8 above)
Comment (tnozicka) [reply pkg/controller/nodeconfig/sync_roles.go:23]: You already give it a set of claimed and pre-existing objects so it's assumed you own them by how you've claimed them in the fist place. If you were to do a patch or update instead you'd not have rechecked again. For the case of the helper it's cheap but I'd not couple it to this PR.

File: pkg/resourceapply/rbac_test.go (hunk 7 above)
Comment (tnozicka) [reply pkg/resourceapply/rbac_test.go]: fixed, thx

File: pkg/controller/nodeconfig/sync.go (hunk 1 above)
Comment (tnozicka) [reply pkg/controller/nodeconfig/sync.go:81]: updated based on #2000

File: pkg/controller/nodeconfig/sync_roles.go [hunk 9]
```diff
@@ -0,0 +1,47 @@
...
+		ctx,
+		requiredRoles,
+		roles,
+		&controllerhelpers.PruneControlFuncs{
+			DeleteFunc: ncc.kubeClient.RbacV1().Roles(nc.Namespace).Delete,
+		},
+		ncc.eventRecorder)
+	if err != nil {
+		return fmt.Errorf("can't prune RoleBinding(s): %w", err)
```
Comment (zimnx) [inline pkg/controller/nodeconfig/sync_roles.go:32]:
```suggestion
		return fmt.Errorf("can't prune Role(s): %w", err)
```

File: pkg/controller/nodeconfig/sync_roles.go (hunk 8 above)
Comment (zimnx) [reply pkg/controller/nodeconfig/sync_roles.go:23]: I meant to add a safety check so we won't delete objects we weren't meant to delete. Filing an issue for future improvement is fine.

Comment (tnozicka) [reply pkg/controller/nodeconfig/sync_roles.go:23]: https://github.com/scylladb/scylla-operator/issues/2010

File: pkg/controller/nodeconfig/sync_roles.go (hunk 9 above)
Comment (tnozicka) [reply pkg/controller/nodeconfig/sync_roles.go]: fixed, thx

**Comment by tnozicka** [comment]
https://github.com/scylladb/scylla-operator/issues/1996#issuecomment-2216892629
/retest

**Comment by tnozicka** [comment]
https://github.com/scylladb/scylla-operator/issues/1996#issuecomment-2217136797
/retest

**Pull request merged**